library/core/src/hint.rs

```diff
@@ -98,9 +98,13 @@ use crate::intrinsics;
 #[rustc_const_stable(feature = "const_unreachable_unchecked", since = "1.57.0")]
 #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
 pub const unsafe fn unreachable_unchecked() -> ! {
-    // SAFETY: the safety contract for `intrinsics::unreachable` must
-    // be upheld by the caller.
-    unsafe { intrinsics::unreachable() }
+    if cfg!(debug_assertions) {
```

Comment:
Should this use [`assert_unsafe_precondition`](https://github.com/rust-lang/rust/blob/9f82651a5fa4b1d96f55ce5507dd2aa204c7fb61/library/core/src/intrinsics.rs#L2188-L2234) instead?